File: lab-07.py
def is_safe(board, row, col, n):
    # Check left side of the row
    for i in range(col):
        if board[row][i] == 1:
            return False

    # Check upper diagonal on left side
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False

    # Check lower diagonal on left side
    for i, j in zip(range(row, n, 1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False

    return True


def solve_n_queens_util(board, col, n):
    # If all queens are placed, return true
    if col >= n:
        return True

    # Consider this column and try placing this queen in all rows one by one
    for i in range(n):
        if is_safe(board, i, col, n):
            # Place this queen in board[i][col]
            board[i][col] = 1

            # Recur to place rest of the queens
            if solve_n_queens_util(board, col + 1, n):
                return True

            # If placing queen in board[i][col] doesn't lead to a solution, then backtrack
            board[i][col] = 0

    # If queen can't be placed in any row in this column col, then return false
    return False


def solve_n_queens(n):
    # Initialize the board with zeros
    board = [[0] * n for _ in range(n)]

    if not solve_n_queens_util(board, 0, n):
        print("Solution does not exist")
        return False

    # Print the solution
    print_solution(board)
    return True


def print_solution(board):
    for row in board:
        print(" ".join(map(str, row)))


# Input section to take the value of n from the user
n = int(input("Enter the value of n: "))
solve_n_queens(n)

File: Lab(ICE-2202)CPP/lab-07.cpp
#include <iostream>
#include <vector>
using namespace std;

void print_solution(vector<vector<int>> &board, int n)
{
    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < n; ++j)
        {
            cout << board[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}

bool is_safe(vector<vector<int>> &board, int row, int col, int n)
{
    // Check left side of the row
    for (int i = 0; i < col; ++i)
    {
        if (board[row][i] == 1)
        {
            return false;
        }
    }

    // Check upper diagonal on left side
    for (int i = row, j = col; i >= 0 && j >= 0; --i, --j)
    {
        if (board[i][j] == 1)
        {
            return false;
        }
    }

    // Check lower diagonal on left side
    for (int i = row, j = col; i < n && j >= 0; ++i, --j)
    {
        if (board[i][j] == 1)
        {
            return false;
        }
    }

    return true;
}

bool solve_n_queens_util(vector<vector<int>> &board, int col, int n)
{
    if (col >= n)
    {
        print_solution(board, n);
        return true;
    }

    bool res = false;
    for (int i = 0; i < n; ++i)
    {
        if (is_safe(board, i, col, n))
        {
            board[i][col] = 1;

            res = solve_n_queens_util(board, col + 1, n) || res;

            board[i][col] = 0; // Backtrack
        }
    }

    return res;
}

void solve_n_queens(int n)
{
    vector<vector<int>> board(n, vector<int>(n, 0));

    if (!solve_n_queens_util(board, 0, n))
    {
        cout << "Solution does not exist." << endl;
    }
}

int main()
{
    int n;
    cout << "Enter the number of queens (N): ";
    cin >> n;

    solve_n_queens(n);

    return 0;
}
